docs/source/main_classes/trainer.rst

```diff
@@ -96,3 +96,216 @@ TFTrainingArguments
+fit bigger models.
+
+The main part that is being integrated at the moment is based on the paper `ZeRO: Memory Optimizations Toward Training
+Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, Olatunji Ruwase, Yuxiong He
+<https://arxiv.org/abs/1910.02054>`__.
+
+These parts are supported by FairScale and DeepSpeed that have been integrated into ``transformers``' Trainer.
+
+FairScale
+-----------------------------------------------------------------------------------------------------------------------
+
+You can already deploy the following features from this paper:
+
+* Optimizer State Sharding
+* Gradient Sharding
+
+using the `--sharded_ddp` trainer argument. This is implemented via `fairscale
+<https://github.com/facebookresearch/fairscale/>`__, so you will have to install this library.
+
+This feature requires distributed training (so multiple GPUs) and is not implemented for TPUs.
+
+For example here is how you could use it for ``finetune_trainer.py``:
+
+.. code-block:: bash
+
+    cd examples/seq2seq
+    python -m torch.distributed.launch --nproc_per_node=2 ./finetune_trainer.py \
+    --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
+    --output_dir output_dir --overwrite_output_dir \
+    --do_train --n_train 500 --num_train_epochs 1 \
+    --per_device_train_batch_size 1  --freeze_embeds \
+    --src_lang en_XX --tgt_lang ro_RO --task translation \
+    --fp16 --sharded_ddp
+
+Note that it works with `--fp16` too, to make things even faster.
+
+One of the main benefits of enabling `--sharded_ddp` is that it uses a lot less GPU memory, so you should be able to
+use significantly larger batch sizes using the same hardware (e.g. 3x or bigger).
+
+DeepSpeed
+-----------------------------------------------------------------------------------------------------------------------
+
+The other important third party component that this trainer supports is `DeepSpeed
+<https://github.com/microsoft/DeepSpeed>`__.
+
+It implements almost everything described in the `ZeRO paper <https://arxiv.org/abs/1910.02054>`__. As of this writing
+it is still missing ZeRO's stage 3. "Parameter Partitioning (Pos+g+p )", but it fully supports:
+
+1. Optimizer State Partitioning (stage 1)
+2. Add Gradient Partitioning (stage 2)
+
+To enable DeepSpeed you need to first install DeepSpeed following `the instructions
+<https://github.com/microsoft/deepspeed#installation>`__.
+
+And when the installation has been completed you need to adjust the command line arguments as following:
+
+1. replace ``python -m torch.distributed.launch`` with ``deepspeed``
+2. add a new argument ``--deepspeed ds_config.json`` where ``ds_config.json`` is the DeepSpeed configuration file as
+   documented at https://www.deepspeed.ai/docs/config-json/
+
+Therefore if your original program looked as following:
+
+.. code-block:: bash
+
+    python -m torch.distributed.launch --nproc_per_node=2 your_program.py <your program\'s normal args>
```

Comment:
I think you should discuss how to configure multi-gpu and multi-node runs with deepspeed. For example, what replaces --nproc_per_node when using deepspeed. A good pointer would be [this](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node).

Reply:
That's super useful - thank you for the suggestion and the link to the resource, @tjruwase. 

I updated the doc.

I'm sure the doc will evolve as users will start asking questions.

p.s for some reason github doesn't ever show your username in the @ completion, even if I type half the first letters of it - very odd.